scripts/broadcast-nmdp-discover.nse

```diff
@@ -0,0 +1,231 @@
+  listener:set_timeout(500)
+  listener:pcap_open(interface.device, 1024, true, filter)
+
+  while (nmap.clock_ms() - start) < timeout do
+    status, _, _, l3data = listener:pcap_receive()
+
+    if not status then
+      goto continue
+    end
+
+    local p = packet.Packet:new(l3data, #l3data)
+    local data = l3data:sub(p.udp_offset + 9)
+
+    if #data < 4 then
+      goto continue
+    end
+
+    stdnse.print_debug(1, "Received NMDP response from %s (%s bytes)", p.ip_src, string.len(data))
+
+    local tlv_type, tlv_len, tlv_value, pos
+    pos = 1
+    --local header = data:sub(pos, pos + 1)
+    pos = pos + 2
+    --local seqno = data:sub(pos, pos + 1)
+    pos = pos + 2
+
+    local mac_address, identity, version, platform, uptime, software_id, board, unpacking, interface
+    while (pos < #data) do
+      -- TLV Type - Unsigned integer (2 bytes)
+      tlv_type = string.unpack(">I2", data:sub(pos, pos + 1))
+      pos = pos + 2
+
+      -- TLV Length - Unsigned integer (2 bytes)
+      tlv_len  = string.unpack(">I2", data:sub(pos, pos + 1))
+      pos = pos + 2
+
+      -- TLV Value
+      tlv_value = data:sub(pos, pos + tlv_len - 1)
+      pos = pos + tlv_len
+
+      --stdnse.print_debug(2, "TLV Type: %s", tlv_type)
+      --stdnse.print_debug(2, "TLV Length: %s", tlv_len)
+      --stdnse.print_debug(2, "TLV Value: %s", stdnse.tohex(tlv_value))
+
+      -- MAC address
+      if tlv_type == 0x01 then
+        mac_address = stdnse.format_mac(tlv_value)
+
+      -- Identity
+      elseif tlv_type == 0x05 then
+        identity = tlv_value
+
+      -- Version
+      elseif tlv_type == 0x07 then
+        version = tlv_value
+
+      -- Platform
+      elseif tlv_type == 0x08 then
+        platform = tlv_value
+
+      -- Uptime - unsigned integer (big endian)
+      elseif tlv_type == 0x0a then
+        uptime_num = string.unpack("<I", tlv_value)
+        local h = math.floor(uptime_num / 3600)
+        local m = math.floor((uptime_num - (h * 3600)) / 60)
+        uptime = h .. "h" .. m .. "m"
+
+      -- Software ID
+      elseif tlv_type == 0x0b then
+        software_id = tlv_value
+
+      -- Board
+      elseif tlv_type == 0x0c then
+        board = tlv_value
+
+      -- Unpacking - unsigned integer
+      elseif tlv_type == 0x0e then
+        if tlv_value:byte(1) == 0x01 then
+          unpacking = "None"
+        else
+          unpacking = "Unknown"
+        end
+
+      -- Interface
+      elseif tlv_type == 0x10 then
+        interface = tlv_value
+
+      -- Unknown
+      else
+        stdnse.print_debug(2, "Unknown TLV Type: %s", tlv_type)
+      end
+    end
+
+    local str = ("MAC Address: %s, IP Address: %s; Identity: %s; Version: %s; Platform: %s; Software ID: %s; Uptime: %s; Board: %s; Unpacking: %s; Interface: %s"):format(
+      mac_address, p.ip_src, identity, version, platform, software_id, uptime, board, unpacking, interface)
```

Comment:
Try this instead to keep the same output but add structured XML output, too:

```lua
local str = stdnse.output_table() -- preserve key order
str["MAC Address"] = mac_address
str["IP Address"] = p.ip_src
str["Identity"] = identity
-- etc.
local mt = getmetatable(str) or {}
mt.__tostring = function(self)
  local t = {}
  for k, v in pairs(self) do
    t[#t+1] = ("%s: %s"):format(k, v)
  end
  return table.concat(t, "; ")
end
setmetatable(str, mt)
```

Reply:
Done